[WIP] Discard whitespace after potential hint trigger matching.

oblitum opened the pull request:
This is useful and noticeable with new kinds of semantic completion
like argument hints. For example, for when you want to pass several
arguments to a function, each argument in its own line, and still want
to have hints for each argument across lines.

This relax the first phase of trigger verification to allow semantic
triggering to be possible after `[`, `(`, `,`, `:` (with any whitespace
following), in the last phase (see hint trigger support in [1](https://github.com/Valloric/ycmd/pull/305)). This is
hard-coded for now since client has no access to actual per-file trigger
matching functions.
- Side effects:

The user will be unable to enter TAB after a hint trigger in case TAB is
listed in `g:ycm_key_list_select_completion`, which it is by default.
This already happens, but now this behavior extends for as much
consecutive whitespace exists after the hint trigger, including line breaks.

It just affect users of argument hints that want to put tabulation for
arguments expanding across lines.

For best argument hints usage it's advised to remove TAB from
`g:ycm_key_list_select_completion` and instead rely on standard Vim
hotkeys (CTRL-N and CTRL-P) for popup menu selection (no need to list
them in `g:ycm_key_list_select_completion`, since they're standard Vim
shortcuts).

This activates issue #1352, but only after `[`, `(`, `,`, `:` (with any
whitespace following).

Check:
- https://github.com/Valloric/YouCompleteMe/pull/1337#issuecomment-71755134

for further arguments of this approach over other possible ones.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/1902)

<!-- Reviewable:end -->

oblitum commented:
Checking tests now... but discussion may start :-)

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7tRij7hrQD0UnNhIgN:-2039421876)*
<!-- Sent from Reviewable.io -->

oblitum commented:
OK, passing. These last two pull requests are the final pieces in the hints jigsaw (at last as I've explained in #1300).

- Hability for the completer to know current trigger and classify it.
- Empty insertion text for pure display popups.
- Whitespace agnostic completion, for per-param hints in multiline calls.
- Some refactoring.

After this, it's a matter completers of taking advantage, as I'll propose for C/C++ and Swift.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7tde0xsdnY9-Nt5WbX:-715228436)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Hm, not sure I'm a fan of this change. Seems like it might have a perf impact. We do more stuff on each keystroke and we do it more often.

Also, the `*Snapped` function needs _very_ good test coverage, at a minimum.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7xwrsxmXuZmXMddndj:-542339966)*
<!-- Sent from Reviewable.io -->

oblitum commented:
As you say, perf impacts? show me the stats. =) I don't see perf impacts both logically, as by daily use. I may be inclined to think how there could be a perf impact, a worst scenery  where we typed using large amout of whitespace between characters frequently, which is not the case, the while loop should barely loop, if ever, most of the time and the identifier check of a single char should be cheap.

I agree with test coverage, will try to add them.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7xyiF6ay_PDdVL4zXx:-201063267)*
<!-- Sent from Reviewable.io -->

Valloric commented:
> As you say, perf impacts? show me the stats. =) 

Well said and fair point. :)

But let's make sure we understand what extra work this would be doing for every keystroke. Not everyone is running a beefy machine like me (and probably you). 

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7y-CP-yNPqUdqXXAsG:-1199108639)*
<!-- Sent from Reviewable.io -->

oblitum commented:
Maybe the addition of this logic gets balanced by the remotion of the emulated TextChangedI, if we get TextChangedI in anytime ;-)

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K7y17z1N_exyF6JW5V5:-392692666)*
<!-- Sent from Reviewable.io -->

davits commented:
You should be able to workaround "Side effect" by checking what kind of items are contained in popup menu.
i.e. `inoremap <expr> <Tab> * pumvisible() && !argument_hints "\<C-n>" : "\<Tab>"`

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-:-K8-W3XaHvt3_llFAs5t:1287657266)*
<!-- Sent from Reviewable.io -->

oblitum commented:
For record: this augments #1352.

homu commented:
:umbrella: The latest upstream changes (presumably #1917) made this pull request unmergeable. Please resolve the merge conflicts.

oblitum commented:
I've changed the intent of this pull request quite a bit given the conflicts I was having with the original one regarding Vim's auto indentation behaviour. I've tried to follow my own recipe for well behaving Vim completion [[1][]] and has possibly reached a good enough solution for my intent. Originally from YCM side I could only make use of `identifier_utils.IsIdentifier` to guess a potential trigger (there were no access to trigger matching from client side), and allow for a whitespace agnostic sematic triggering. In this case now I'm just assuming a limited number of potential hint triggers for that.

Auto indent conflicts as mentioned in #1352 can't simply be evaded the way YCM/ycmd works now when whishing to offer whitespace agnostic sematic completion. Because it's going to break bullet 2 of [[1][]]. I'm doing this consciously in this set of changes, but just in narrowed occasions, which in
my view, provides a good balance.

This means that this activates #1352, but only after `[`, `(`, `,`, `:`, plus whitespace.

I'll add tests by next week probably, but from a high level, if anyone got an opinion on this I'd be glad to hear. Sorry for the long hiatus.
 
[1]: https://github.com/Valloric/YouCompleteMe/issues/1352#issuecomment-172135046

oblitum commented:
Guys, can anyone please tell what's the YCM way of doing the following (part of this pull request) to avoid unicode incompatibilities?

[[1](https://github.com/Valloric/YouCompleteMe/pull/1902/commits/ac82899433fa577193ddf2046dcf7ceeaa7b4d79#diff-414ad007afc1163aea017b394fe82f27R71)]: `if str( vim.current.buffer[ line ][ column - 1 ] ) in potential_hint_triggers`

P.S. Dude, unicode and python is a nightmare...

puremourning commented:
why does it currently need the `str` call?

oblitum commented:
@puremourning I don't recall, formally it wasn't there, I added it when I got a bug, long time ago before I rebased, and it fixed it, something related with that line and column not being a char for some reason....

oblitum commented:
I've tried to patch it now with `unicode` instead of `str` (and making trigger list `unicode`), still unicode issues. I guess I'll remove `str`/`unicode` and try to understand the initial bug that I forgot, if it ever comes up. In any case, if there's a better solution than this I'm all ears.

oblitum commented:
Removing `str` and pasting some chinese charaters in my buffer I get a UnicodeWarning that happens only the first time I past it.

```
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
/home/francisco/.vim/plugged/YouCompleteMe/autoload/../python/ycm/vimsupport.py:70: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  if vim.current.buffer[ line ][ column - 1 ] in potential_hint_triggers:
```

puremourning commented:
You probably just need to use `ycmd.utils.ToUnicode` rather than `unicode` cos that won't work on py3

oblitum commented:
Using `ToUnicode` which is already imported in this file, and turning trigger list to unicode chars I get when pasting chinese string into a buffer:

```
Messages maintainer: Bram Moolenaar <Bram@vim.org>
"config.toml" 14L, 487C
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
Traceback (most recent call last):
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "<string>", line 1, in <module>
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/autoload/../python/ycm/youcompleteme.py", line 272, in OnFileReadyToParse
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    self._latest_file_parse_request.Start()
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/autoload/../python/ycm/client/event_notification.py", line 43, in Start
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    request_data = BuildRequestData()
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/autoload/../python/ycm/client/base_request.py", line 158, in BuildRequestData
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    line, column = vimsupport.SnappedLineAndColumn()
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/autoload/../python/ycm/vimsupport.py", line 70, in SnappedLineAndColumn
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    if ToUnicode( vim.current.buffer[ line ][ column - 1 ] ) in potential_hint_triggers:
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/python/ycm/../../third_party/ycmd/ycmd/utils.py", line 89, in ToUnicode
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    return str( value, 'utf8' )
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/home/francisco/.vim/plugged/YouCompleteMe/third_party/ycmd/third_party/python-future/src/future/types/newstr.py", line 97, in __new__
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    value = args[0].decode(*args[1:], **kwargs)
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
  File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
    return codecs.utf_8_decode(input, errors, True)
Error detected while processing function <SNR>207_OnCursorMovedNormalMode[5]..<SNR>207_OnFileReadyToParse:
line   13:
UnicodeDecodeError: 'utf8' codec can't decode byte 0xa1 in position 0: invalid start byte
```

oblitum commented:
This should be probably related the way indexing is done, not sure how to fix it.

oblitum commented:
Ah, I recalled the reason I used `str` formally I guess, it was because, in my mind, even if `vim.current.buffer[ line ][ column - 1 ]` was the last part of a multibyte character, that slice should be enough to check difference with the small and specific trigger list, in case the indexing was indeed getting part of a utf-8 char, otherwise, it would work too with proper conversion of both sides. Anyways, I worked on speculation here.

oblitum commented:
I guess, following this line of reason, if indexing is indeed getting the last byte of the unicode char, this part may be indeed a `invalid start byte` for utf-8, just like the error says when trying the conversion to unicode.

puremourning commented:
It all depends on whether `column` is a byte offset or a "character" offset. If it is a byte offset, and you want the previous "character", then you probably need to use the new `ByteOffsetToCodepointOffset` from `ycmd.request_wrap`. If it is a character offset, then you just need to convert the line value to unicode first, like: `ToUnicode( vim.current.buffer[ line ] )[ column - 1]`.

The problem is that with byte offsets, you can't use `- 1` to get the previous "character". If you just want to compare the raw bytes, then use `ToBytes()` on the Right Hand Side, like `vim.current.buffer[ line ][ column - 1] == ToBytes( the_other_thing )`

oblitum commented:
ah, I guess I now get what's happening.... this chinese character last byte, it can't be string, because 0xa1 > 127, but it can't be utf-8 because it's invalid utf-8 start.

oblitum commented:
I'll just convert everything to bytes, wtf.

oblitum commented:
@puremourning yes indeed, thanks for the overall analysis from all sides!

puremourning commented:
Actually, you need to use `ToBytes` on both sides, so that it works the same in PY3

puremourning commented:
> P.S. Dude, unicode and python is a nightmare...

https://www.google.co.uk/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&ved=0ahUKEwi3xv22q6jMAhXENhoKHUjHDmgQjRwIBw&url=%2Furl%3Fsa%3Di%26rct%3Dj%26q%3D%26esrc%3Ds%26source%3Dimages%26cd%3D%26cad%3Drja%26uact%3D8%26ved%3D0ahUKEwi3xv22q6jMAhXENhoKHUjHDmgQjRwIBw%26url%3Dhttp%253A%252F%252Fwww.memecenter.com%252Ffun%252F750815%252Foc---no-fucking-shit%26psig%3DAFQjCNG0WR5zja5R7BZvJazl6vtvpYuMBg%26ust%3D1461623868632758&psig=AFQjCNG0WR5zja5R7BZvJazl6vtvpYuMBg&ust=1461623868632758

oblitum commented:
@puremourning I'm reading about this (http://stackoverflow.com/questions/6269765/what-does-the-b-character-do-in-front-of-a-string-literal) because I want just to use byte literals. I've tried here, and now it's working, I've removed `str`/`unicode` from left side, and no conversion on right side.

So I guess, as you state, putting ToBytes on left side should be enough.

puremourning commented:
The `b` literal only sorta works, in the sense that we use Python-future which does unspeakable things to python 2 standard library and stuff. Technically under python-future, we should use `bytes( b'blah blah' )`, but that's why we have `ycmd.utils.ToBytes` - it just works on py2 and py3.

I think this is the page that explains it: http://python-future.org/what_else.html#what-else

oblitum commented:
OK, nevermind, I'll just use ToBytes. Thanks for the link, unicode in python is such a BIG realm.

oblitum commented:
Well, no more errors.... but also, no more space agnostic hints!....

Now will have to figure out why comparison of `ToBytes` is failing, but str isn't.

oblitum commented:
Ah ok... iterators trap.

vheon commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/ycm/vimsupport.py, line 71 \[r4\]](https://reviewable.io:443/reviews/valloric/youcompleteme/1902#-KGEEZizVUq5nVBaG3eS:-KGEEZizVUq5nVBaG3eT:-2048752306) ([raw file](https://github.com/valloric/youcompleteme/blob/214b2f0ef530e10b521099156ec5395875228781/python/ycm/vimsupport.py#L71)):*
After the work that @puremourning did on `ycmd` about codepoints vs bytes this `column - 1` really scares me :confused: What are we talking about here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/youcompleteme/1902)*
<!-- Sent from Reviewable.io -->

oblitum commented:
@vheon I think all possible issues regarding Unicode handling got fixed by now. `vim.current.buffer[ line ][ column - 1 ]` is the last byte of a character (`unicode` or `char`) that's not whitespace, before caret.

This byte is enough to compare with the single-byte characters for the possible hint triggers listed. IF the last byte is part of some random Unicode character that happens to be equal to a trigger (dunno whether Unicode allows this), stills, it won't be an issue, because such condition of coding identifiers in Unicode is such a rare situation, if allowed, and still, it would just trigger for no results in the end.

oblitum commented:
Checking the design of UTF-8 code units, the last byte of multibyte unicode character can never be in the range < 127. So, no issues regarding my last speculation.

oblitum commented:
ah, sorry my bad, I've just deleted my previous comment to leave things cleaner but my ending conclusion that the last byte of a multi-byte Unicode character not being possible to be < 127 was flat wrong since just the ending 6 bits are used to represent those bytes, it doesn't include the leading `10`'s of full continuation byte...

So, again, I'll now illustrate the problem situation I was speculating with a practical example.

```
28 == (
E2 28 == â
```

So, in this case, if the guy types an `â`, the last byte will be 28, the same as that of a hint trigger and it's going to try semantic completion, but will not return results.

I said that, as it's such a rarity to code in Unicode outside comments, that this would not be a real issue. It won't blow up in case it happens, and it should almost never happen anyway.

oblitum commented:
Awwww, consider that example as "theorical", the values are wrong, but the problem situation is correctly presented. Haha, the colors on the Wikipedia page made me doubt my first conclusion that was indeed correct, the bytes are full continuation bytes, just not the ending 6 bits, my bad again.

So, what I described, is a non-issue, and just can't happen. Sorry for the noise, I think I finally came to read all important parts of the UTF-8 spec b/c of this.

homu commented:
:umbrella: The latest upstream changes (presumably #2223) made this pull request unmergeable. Please resolve the merge conflicts.

bstaletic commented:
What's the status of this PR?
As it stands now, this PR exhibits oblitum/YouCompleteMe#18, which has been fixed in @oblitum's fork.

oblitum commented:
Sorry, I don't think I'm going contribute to YCM anymore.
